Tuple type introduction and about added.

rajeebkm opened the pull request:
Closes #40 

- Tuple type introduction and about added.

misicnenad commented:
Please resolve the merge conflicts

rajeebkm commented:
> Please resolve the merge conflicts

Resolved the merge conflicts. but getting markdown linting issues. Tried to fix, as per suggestion, nut it seems still some issues with linting presents. Could you please help  with this ?

misicnenad commented on concepts/booleans/.meta/config.json:
```diff
@@ -1,8 +1,7 @@
 {
-    "blurb": "As in most other programming languages, a boolean type in Cairo has two possible values: `true` and `false`.",
+    "blurb": "As in most other programming languages, a Boolean type in Cairo has two possible values: true and false.",
```
Remove all changes to:
- booleans
- felts
- integers

We'll address some of these in a later PR.

misicnenad commented on concepts/tuples/about.md:
```diff
@@ -1 +1,53 @@
+
+## Creating Tuples
+
+To create a tuple, you write a comma-separated list of values inside parentheses. Each position in the tuple has a specific type, and these types can differ from one another. Here’s an example with optional type annotations:
```
```suggestion
To create a tuple, you write a comma-separated list of values inside parentheses. Each position in the tuple has a specific type, and these types can differ from one another.
```
nit: redundant

misicnenad commented on concepts/tuples/about.md:
````diff
@@ -1 +1,53 @@
+}
+```
+
+In this example, the variable `tup` binds to the entire tuple, treating it as a single compound element.
````
```suggestion
```
Already stated this in the first sentence.
Instead, state that a tuple can hold any other data type, even other tuples, and give a brief example.

misicnenad commented on concepts/tuples/about.md:
```diff
@@ -1 +1,53 @@
+
+    let (x, y, z) = tup;
+
+    if y == 6 {
```
```suggestion
```
Instead of the if-statement, print each tuple field like: 
```rust
println!("x is: {x}"); // prints "x is: 500"
```

misicnenad commented on concepts/tuples/about.md:
````diff
@@ -1 +1,53 @@
+}
+```
+
+This code creates a tuple and binds it to the variable `tup`. Using a `let` pattern, it destructures `tup` into three separate variables: `x`, `y`, and `z`. The program then checks if `y` is 6 and prints a corresponding message.
````
```suggestion
```
No need for an explanation after each example, unless the example does something really complex or counterintuitive.

misicnenad commented on concepts/tuples/about.md:
```diff
@@ -1 +1,53 @@
+
+## Declaring and Destructuring Simultaneously
+
+You can also declare and destructure a tuple simultaneously:
```
```suggestion
You can also declare and destructure a tuple simultaneously.
```
nit

misicnenad commented on concepts/tuples/about.md:
```diff
@@ -1 +1,53 @@
+
+## Destructuring Tuples
+
+To access the individual values within a tuple, you can use pattern matching to destructure the tuple. This breaks the tuple into its constituent parts. For example:
```
```suggestion
To access the individual values within a tuple, you can use pattern matching to destructure the tuple. This breaks the tuple into its constituent parts.
```
nit

misicnenad commented on concepts/tuples/about.md:
```diff
@@ -1 +1,53 @@
+
+In Cairo, every expression returns a value, and when an expression returns nothing, it implicitly returns `()`. This can be useful for functions that need to return a value but have no meaningful value to return.
+
+## Summary
```
```suggestion
## Named Tuple Type Declaration
```
No need for the summary. Instead, give an example of declaring a named tuple type using the keyword `type`, like:
```rust
type Point = (u32, u32);

fn main() {
    let point: Point = (10, 8);
    println!("{point:?}"); // prints "(10, 8)"
}
```

misicnenad commented on concepts/tuples/introduction.md:
```diff
@@ -1 +1,16 @@
 # Introduction
+
+Tuples are a fundamental compound data type in Cairo, providing an efficient way to group related data of different types into a single, fixed-length structure. This allows you to manage and access related data seamlessly. Here are a couple of examples to illustrate their usage:
```
The introduction should be short, and ideally shouldn't re-display examples from _about.md_.

Instead, you can reuse Rust's Tuple intro:
> Tuples are a lightweight way to group a fixed set of arbitrary types of data together. Tuples can have an arbitrary number of elements. You can access the elements by destructuring.

misicnenad reviewed (changes requested):
Short and sweet :) Just some details to address

misicnenad commented on concepts/tuples/.meta/config.json:
```diff
@@ -1,7 +1,7 @@
-        "misicnenad"
+        "rajeebkm"
     ],
     "contributors": []
```
List me as a contributor

rajeebkm commented:
> > Please resolve the merge conflicts
> 
> Resolved the merge conflicts, but getting markdown linting issues. Tried to fix, as per suggestion, but it seems still some issues with linting presents. Could you please help with this ?

Hi @misicnenad I resolved the megre conflicts and linting issue. Please check once. Thanks

misicnenad commented on concepts/tuples/about.md:
````diff
@@ -4,37 +4,33 @@ A tuple is a versatile and efficient compound data type in Cairo, allowing you t
 }
 ```
 
-In this example, the variable `tup` binds to the entire tuple, treating it as a single compound element.
````
Don't forget to address [this change request](https://github.com/exercism/cairo/pull/106#discussion_r1700270313):
> Instead, state that a tuple can hold any other data type, even other tuples, and give a brief example.

misicnenad commented on concepts/tuples/about.md:
```diff
@@ -4,37 +4,33 @@ A tuple is a versatile and efficient compound data type in Cairo, allowing you t
-    if y == 6 {
-        println!("y is 6!");
-    }
+    println!("x is: ", {x}); // prints "x is: 500"
```
```suggestion
    println!("x is: {x}"); // prints "x is: 500"
```
update the other printlns the same way, the current code will not compile due to wrong syntax

misicnenad commented on concepts/tuples/about.md:
```diff
@@ -48,6 +44,13 @@ The unit type is a special type in Cairo that has only one value: `()`. It is re
 
-## Summary
+## Named Tuple Type Declaration
+
```
Add a short explanation of the named tuple type syntax

rajeebkm commented:
Hi @0xNeshi I have addressed the issues. Please review.

0xNeshi commented on concepts/tuples/.meta/config.json:
```diff
@@ -1,7 +1,8 @@
     "authors": [
-        "rajeebkm"
+        "rajeebkm",
+        "0xNeshi"
```
```suggestion
```
You're the sole author :) If nothing else, you can add me as a contributor for reviewing the PR

0xNeshi commented on concepts/booleans/about.md:
```diff
@@ -1,3 +1,4 @@
+
```
Don't forget to revert all changes that are not tuple related, see https://github.com/exercism/cairo/pull/106#discussion_r1700266710

0xNeshi replied:
```suggestion
```

rajeebkm commented:
Hi @0xNeshi I have resolved the issues and reverted the changes as requested.

0xNeshi reviewed (changes requested):
The changes to unrelated concepts are still present :( 
Other than that, the PR looks great, so we can merge as soon as you remove the former :)

rajeebkm commented:
> The changes to unrelated concepts are still present :( Other than that, the PR looks great, so we can merge as soon as you remove the former :)

I believe I have reverted and modified the changes. Could you point me out the eaxt unrelated concepts that are still present ?
Thanks!

0xNeshi commented:
> I believe I have reverted and modified the changes. Could you point me out the eaxt unrelated concepts that are still present ? Thanks!

![image](https://github.com/user-attachments/assets/027b19ff-dc57-4c2f-bd82-804e56eeb9d5)

rajeebkm commented:
> > I believe I have reverted and modified the changes. Could you point me out the eaxt unrelated concepts that are still present ? Thanks!
> 
> You can see the affected files in the "files" tab ![image](https://private-user-images.githubusercontent.com/19427053/360858973-027b19ff-dc57-4c2f-bd82-804e56eeb9d5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ0OTg5MDcsIm5iZiI6MTcyNDQ5ODYwNywicGF0aCI6Ii8xOTQyNzA1My8zNjA4NTg5NzMtMDI3YjE5ZmYtZGM1Ny00YzJmLWJkODItODA0ZTU2ZWViOWQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI0VDExMjMyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUxNzE2MTRkYzgwYzhhMDM1ODY3ZjUyZjAwYWJmMGY5NWY3MTJiYjU1ZjUyYTFmZTAyNGEzOWVlY2EwYmNkMmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ltMtjPMwkIPk0TSIOevls3yWqdpdDaoYMHNih6LidhE)

It showing as affected as I have resolved the linting issues. Concepts are as it is before tuple push I believe.

0xNeshi commented:
> It showing as affected as I have resolved the linting issues. Concepts are as it is before tuple push I believe.

They're not, you can open the "files" tab and see some leftover non-lint changes

0xNeshi commented on concepts/booleans/.meta/config.json:
```diff
@@ -1,5 +1,5 @@
 {
-    "blurb": "As in most other programming languages, a boolean type in Cairo has two possible values: `true` and `false`.",
+    "blurb": "As in most other programming languages, a Boolean type in Cairo has two possible values: true and false.",
```
```suggestion
    "blurb": "As in most other programming languages, a boolean type in Cairo has two possible values: `true` and `false`.",
```

0xNeshi commented on concepts/booleans/about.md:
```diff
@@ -23,7 +24,7 @@ One important aspect to note is that when declaring a boolean variable in Cairo,
 ## Example Code
 
-The following code snippet demonstrates how to use boolean types in Cairo. It shows the declaration of boolean variables, assignment of boolean expressions, and usage of assertions to verify the correctness of the boolean logic:
+The following code snippet demonstrates how to use boolean types in Cairo. It shows the declaration of boolean variables, assignment of boolean expressions, and usage of assertions to verify the correctness of the boolean logic.
```
```suggestion
The following code snippet demonstrates how to use boolean types in Cairo. It shows the declaration of boolean variables, assignment of boolean expressions, and usage of assertions to verify the correctness of the boolean logic:
```

0xNeshi commented on concepts/booleans/introduction.md:
```diff
@@ -1,4 +1,6 @@
+
```
```suggestion
```

0xNeshi commented on concepts/booleans/introduction.md:
```diff
@@ -1,4 +1,6 @@
 # Introduction
 
 In Cairo, similar to most other programming languages, the Boolean type represents a binary state with two possible values: `true` and `false`.
+
```
```suggestion
```

0xNeshi commented on concepts/felts/about.md:
```diff
@@ -17,6 +17,7 @@ $\frac{x}{y} \cdot y = x$
 
 - If $y$ divides $x$ as integers, Cairo will provide the expected result. For instance, $\frac{6}{2}$ will yield $3$.
-- If $y$ does not divide $x$ , the result can be surprising. For example, since $2 \cdot \left(\frac{P+1}{2}\right) = P + 1 \equiv 1 \mod P$, the value of $\frac{1}{2}$ in Cairo is $\frac{P + 1}{2}$ (not $0$ or $0.5$), as it satisfies the above equation.
+
```
```suggestion
```

0xNeshi commented on concepts/integers/about.md:
```diff
@@ -13,47 +13,6 @@ Cairo also provides support for signed integers, starting with the prefix `i`. T
 | 64-bit   | u64      | i64      |
 | 128-bit  | u128     | i128     |
-| 256-bit  | u256     | ------   |
+| 256-bit  | u256     | i256     |
```
```suggestion
| 256-bit  | u256     | ------     |
```

0xNeshi commented on concepts/integers/about.md:
```diff
@@ -13,47 +13,6 @@ Cairo also provides support for signed integers, starting with the prefix `i`. T
-| 256-bit  | u256     | ------   |
+| 256-bit  | u256     | i256     |
 
 There is also a special integer type called `usize`, which is currently just an alias for `u32`. This type might play a more distinct role in the future, once Cairo can be compiled to MLIR (Multi-Level Intermediate Representation).
```
```suggestion
There is also a special integer type called `usize`, which is currently just an alias for `u32`. This type might play a more distinct role in the future, once Cairo can be compiled to MLIR (Multi-Level Intermediate Representation).

## Integer Literals in Cairo

You can write integer literals in any of the forms shown in the below table. Note that number literals that can be multiple numeric types allow a type suffix, such as `57_u8`, to designate the type. It is also possible to use a visual separator `_` for number literals, in order to improve code readability.

| Numeric Literals | Example |
| ---------------- | ------- |
| Decimal          | 98222   |
| Hex              | 0xff    |
| Octal            | 0o04321 |
| Binary           | 0b01    |

So how do you know which type of integer to use? Try to estimate the max value your int can have and choose the appropriate size. The primary situation in which you’d use `usize` is when indexing some sort of collection.

## Numeric Operations

Cairo supports the basic mathematical operations you’d expect for all the integer types: addition, subtraction, multiplication, division, and remainder. Integer division truncates toward zero to the nearest integer. The following code shows how you’d use each numeric operation in a `let` statement:

```rust
fn main() {
    // addition
    let sum = 5_u128 + 10_u128;

    // subtraction
    let difference = 95_u128 - 4_u128;

    // multiplication
    let product = 4_u128 * 30_u128;

    // division
    let quotient = 56_u128 / 32_u128; // result is 1
    let quotient = 64_u128 / 32_u128; // result is 2

    // remainder
    let remainder = 43_u128 % 5_u128; // result is 3
}
```

Each expression in these statements uses a mathematical operator and evaluates to a single value, which is then bound to a variable.

List of all operators that Cairo provides can be found [here](https://book.cairo-lang.org/appendix-02-operators-and-symbols.html#operators)
```

0xNeshi commented on concepts/integers/introduction.md:
```diff
@@ -3,7 +3,3 @@
 Integers are whole numbers with no decimal separator.
 
 Integer types can either store only positive numbers (unsigned) or store either positive and negative numbers (signed).
```
```suggestion
Integer types can either store only positive numbers (unsigned) or store either positive and negative numbers (signed).

Examples are -6, 0, 1, 25 and 976.

Note that you can use underscores to improve the readability of longer numbers: `1_000_000`.
```

0xNeshi commented on concepts/integers/about.md:
```diff
@@ -13,7 +13,7 @@ Cairo also provides support for signed integers, starting with the prefix `i`. T
 | 64-bit   | u64      | i64      |
 | 128-bit  | u128     | i128     |
-| 256-bit  | u256     | ------   |
+| 256-bit  | u256     | ------     |
```
```suggestion
| 256-bit  | u256     | ------   |
```

0xNeshi reviewed (approved):
Reverted the non-tuple changes, lgtm now
